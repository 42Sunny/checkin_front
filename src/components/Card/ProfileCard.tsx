import ListIcon from "@mui/icons-material/List";
import React from "react";
import CardContents from "components/Card/CardContents";
import Circle from "components/common/Circle";
import classes from "styles/components/Card/ProfileCard.module.css";
import useUser from "utils/hooks/useUser";
import logo from "assets/42-logo-black.png";

type UtilBoxProps = {
  handleFlip: (e: React.MouseEvent) => void;
};

function UtilBox({ handleFlip }: UtilBoxProps) {
  return (
    <div className={classes["util-box"]}>
      <ListIcon data-testid='flip-button' onClick={handleFlip} />
    </div>
  );
}

type UserCircleProps = {
  state: string | null;
};
function UserCircle({ state }: UserCircleProps) {
  const color = state === "checkOut" ? "green" : "orange";
  return <Circle color={color} />;
}

type ProfileProps = {
  profile: string;
  userId: string;
  state: string | null;
};

function Profile({ profile, userId, state }: ProfileProps) {
  return (
    <div className={classes["profile-wrapper"]}>
      <UserCircle state={state} />
      <img className={classes.profile} src={profile} alt='profile' />
      <h2 className={classes.userId}>{userId}</h2>
    </div>
  );
}
type IProps = {
  handleFlip: (e: React.MouseEvent) => void;
};

function ProfileCard({ handleFlip }: IProps) {
  const {
    user: { id: userId, profile, state },
  } = useUser();

  return (
    <div className={classes.profileCard}>
      <img className={classes.logo} alt='logo' src={logo} />
      <UtilBox handleFlip={handleFlip} />
      <Profile profile={profile} userId={userId} state={state} />
      <CardContents />
    </div>
  );
}

export default ProfileCard;
