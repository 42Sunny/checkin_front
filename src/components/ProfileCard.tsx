import ListIcon from "@mui/icons-material/List";
import React from "react";
import classes from "../styles/components/ProfileCard.module.css";
import useUser from "../utils/hooks/useUser";
import CheckInUi from "./CheckInUi";
import CheckOutUi from "./CheckOutUi";

interface UtilBoxProps {
  handleFlip: (e: React.MouseEvent) => void;
}
const UtilBox: React.FC<UtilBoxProps> = ({ handleFlip }) => (
  <div className={classes["util-box"]}>
    <ListIcon onClick={handleFlip} />
  </div>
);

interface ProfileProps {
  profile: string;
  userId: string;
}
const Profile: React.FC<ProfileProps> = ({ profile, userId }) => (
  <div className={classes["profile-wrapper"]}>
    <img className={classes.profile} src={profile} alt='profile' />
    <h2>{userId}</h2>
  </div>
);
interface IProps {
  handleFlip: (e: React.MouseEvent) => void;
  handleCheckIn: (cardNum: string) => (e: React.FormEvent<HTMLFormElement>) => Promise<boolean>;
  handleCheckOut: () => Promise<void>;
}

const ProfileCard: React.FC<IProps> = ({ handleFlip, handleCheckIn, handleCheckOut }) => {
  const {
    user: { id: userId, profile, state },
  } = useUser();

  return (
    <div className={classes.profileCard}>
      <UtilBox handleFlip={handleFlip} />
      <Profile profile={profile} userId={userId} />
      <hr className={classes.divider} />
      {state === "checkIn" && <CheckOutUi handleCheckOut={handleCheckOut} />}
      {state === "checkOut" && <CheckInUi handleCheckIn={handleCheckIn} />}
    </div>
  );
};

export default ProfileCard;
