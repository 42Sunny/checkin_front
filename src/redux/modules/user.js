import { DEFAULT_PROFILE } from '../../utils/utils';

// actions
const LOGIN = 'LOGIN';
const LOGOUT = 'LOGOUT';
const SET_USER = 'SET_USER';
const SET_CARDNUM = 'SET_CARDNUM';

// action creators
export const login = () => {
  return {
    type: LOGIN
  };
};

export const logout = () => {
  return {
    type: LOGOUT
  };
};

export const setUser = user => {
  return {
    type: SET_USER,
    payload: user
  };
};

export const setCardNum = cardNum => {
  return {
    type: SET_CARDNUM,
    payload: cardNum
  };
};

// initalState
const initalState = {
  isLogin: false,
  id: '',
  cardNum: '',
  status: 'out',
  profile: DEFAULT_PROFILE
};

// reducer
export const user = (state = initalState, action) => {
  switch (action.type) {
    case LOGIN:
      return {
        ...state,
        isLogin: true
      };
    case LOGOUT:
      return {
        ...state,
        isLogin: false
      };
    case SET_USER:
      return {
        ...state,
        id: action.payload.id,
        cardNum: action.payload.cardNum,
        status: action.payload.status,
        profile: action.payload.profile
      };
    case SET_CARDNUM:
      return {
        ...state,
        cardNum: action.payload
      };
    default:
      return state;
  }
};
