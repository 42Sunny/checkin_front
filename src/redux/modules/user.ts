import { createAction, ActionType } from "typesafe-actions";
import * as Sentry from "@sentry/react";
import DEFAULT_PROFILE from "../../assets/user-default.png";
// actions
const LOGIN = "user/LOGIN";
const LOGOUT = "user/LOGOUT";
const SET_USER = "user/SET_USER";
const SET_CARD_NUM = "user/SET_CARD_NUM";

// action creators
export const login = createAction(LOGIN)();
export const logout = createAction(LOGOUT)();
export const setUser = createAction(SET_USER)<Omit<User, "isLogin" | "isAdmin">>();
export const setCardNum = createAction(SET_CARD_NUM)<{ cardNum: string }>();

// type
const actions = { setCardNum, setUser, login, logout };
type UserActions = ActionType<typeof actions>;

// initialState
const initialState: User = {
  isLogin: false,
  id: "",
  cardNum: "",
  state: "checkOut",
  checkinAt: null,
  checkoutAt: null,
  profile: DEFAULT_PROFILE,
  isAdmin: false,
};

// reducer
const user = (state = initialState, action: UserActions) => {
  switch (action.type) {
    case LOGIN:
      return {
        ...state,
        isLogin: true,
      };
    case LOGOUT:
      return {
        ...state,
        isLogin: false,
      };

    case SET_USER: {
      const { cardNum, checkinAt, checkoutAt, id, profile, state: userState } = action.payload;
      Sentry.setUser({ username: id });
      return { ...state, id, cardNum, state: userState, checkinAt, checkoutAt, profile };
    }
    case SET_CARD_NUM:
      return {
        ...state,
        cardNum: action.payload.cardNum,
      };
    default:
      return state;
  }
};

export default user;
