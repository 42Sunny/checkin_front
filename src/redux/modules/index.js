import { combineReducers } from 'redux';
import { user } from './user';
import { config } from './config';
import { status } from './status';

const rootReducer = combineReducers({
  user: user,
  config: config,
  status: status
});

export default rootReducer;
