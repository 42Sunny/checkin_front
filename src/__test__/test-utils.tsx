// test-utils.jsx
import { configureStore } from "@reduxjs/toolkit";
import { Queries, render, RenderResult } from "@testing-library/react";
import { createMemoryHistory } from "history";
import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
// Import your own reducer
import renderer from "react-test-renderer";
import appStore from "../redux/configureStore";
import rootReducer, { RootState } from "../redux/modules";
import { initialState as clusterInitState } from "../redux/modules/cluster";
import { initialState as userInitState } from "../redux/modules/user";

const makeCustomStore = ({
  cluster,
  user,
}: {
  cluster?: typeof clusterInitState;
  user?: typeof userInitState;
} = {}) => {
  return {
    clusterReducer: { ...clusterInitState, ...cluster },
    userReducer: { ...userInitState, ...user },
  };
};

const customRender = (
  ui: ReactElement,
  preloadedState: RootState = makeCustomStore(),
  { store = configureStore({ reducer: rootReducer, preloadedState }), ...renderOptions } = {},
  history = createMemoryHistory(),
): RenderResult<Queries, HTMLElement> => {
  const Wrapper: any = ({ children }: { children: ReactElement }) => {
    return (
      <Provider store={store}>
        <Router history={history}>{children}</Router>
      </Provider>
    );
  };
  return render(ui, { wrapper: Wrapper, ...renderOptions });
};

const createWithWrapper = (children: ReactElement) =>
  renderer.create(
    <Provider store={appStore}>
      <Router history={createMemoryHistory()}>{children}</Router>
    </Provider>,
  );

// re-export everything
export * from "@testing-library/react";
// override render method
export { customRender as render, createWithWrapper, makeCustomStore };
