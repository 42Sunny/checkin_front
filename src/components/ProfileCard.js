import { useState, useEffect, useCallback } from 'react';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import { useHistory } from 'react-router-dom';
import Button from '../components/Button';
import Profile from '../components/Profile';
import CheckInForm from '../components/CheckInForm';
import CheckInInfo from '../components/CheckInInfo';
import { checkOut, checkIn } from '../api/api';

import { setCardNum } from '../redux/modules/user';
import AutorenewIcon from '@mui/icons-material/Autorenew';
import '../styles/ProfileCard.css';

const ProfileCard = ({ handleFlip }) => {
  const history = useHistory();
  const { cardNum, status } = useSelector(
    state => ({
      cardNum: state.user.cardNum,
      status: state.user.status
    }),
    shallowEqual
  );

  const dispatch = useDispatch();

  const [checkAll, setCheckAll] = useState(false);
  const [checkStatus, setCheckStatus] = useState([false, false, false]);
  const [readySubmit, setReadySubmit] = useState(false);

  const btnText = status === 'out' ? 'CHECK IN' : 'CHECK OUT';

  const handleCheckIn = useCallback(async () => {
    if (readySubmit) {
      try {
        console.log('cardNum', cardNum);
        await checkIn(cardNum);
        history.push('/end');
      } catch (err) {
        if (err.response.data.code === 404) {
          alert(err.response.data.message);
        } else {
          alert('체크인을 처리할 수 없습니다. 제한 인원 초과가 아닌 경우 관리자에게 문의해주세요.');
        }
        dispatch(
          setCardNum({
            cardNum: ''
          })
        );
      }
    }
  }, [cardNum, readySubmit, history, dispatch]);

  const handleCheckOut = useCallback(async () => {
    if (window.confirm('퇴실 하시겠습니까?')) {
      try {
        await checkOut();
        history.push('/end');
      } catch (err) {
        if (!err.response) {
          alert('정상적으로 처리되지 않았습니다.\n네트워크 연결 상태를 확인해주세요.');
        } else if (err.response.data.code === 404) {
          alert('이미 체크아웃 되었습니다.');
          history.push('/');
        } else {
          alert(err.response.data.message);
        }
      }
    }
  }, [history]);

  const checkSubmitCondition = useCallback(() => {
    if (cardNum && checkAll) {
      setReadySubmit(true);
    } else {
      setReadySubmit(false);
    }
  }, [cardNum, checkAll]);

  useEffect(() => {
    if (status === 'out') {
      checkSubmitCondition();
    }
  }, [cardNum, status, checkSubmitCondition]);

  useEffect(() => {
    if (JSON.stringify(checkStatus) === JSON.stringify([true, true, true])) {
      setCheckAll(true);
    } else {
      setCheckAll(false);
    }
    return () => {
      setCheckAll(false);
    };
  }, [checkStatus]);

  return (
    <div id='profile-card-wrapper'>
      {status === 'in' && (
        <div
          style={{
            textAlign: 'right',
            width: '100%'
            // position: 'fixed',
            // top: '30px',
            // right: '30px'
          }}
        >
          <AutorenewIcon onClick={handleFlip} />
        </div>
      )}
      <Profile />
      {status === 'out' ? (
        <CheckInForm
          checkAll={checkAll}
          setCheckAll={setCheckAll}
          checkStatus={checkStatus}
          setCheckStatus={setCheckStatus}
        />
      ) : (
        <>
          <hr className='divider' />
          <CheckInInfo />
        </>
      )}
      <Button
        className={
          status === 'out' ? `submitBtn out ${readySubmit ? 'ready' : ''}` : 'submitBtn in'
        }
        handleClick={status === 'out' ? handleCheckIn : handleCheckOut}
        text={btnText}
      />
    </div>
  );
};

export default ProfileCard;
