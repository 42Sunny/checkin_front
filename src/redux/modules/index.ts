import { combineReducers } from "redux";
import userReducer from "./user";
import configReducer from "./config";
import statusReducer from "./status";

const rootReducer = combineReducers({
  userReducer,
  configReducer,
  statusReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;
