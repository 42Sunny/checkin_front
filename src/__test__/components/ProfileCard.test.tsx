import userEvent from "@testing-library/user-event";
import ProfileCard from "../../components/ProfileCard";
import { initialState as userInitState } from "../../redux/modules/user";
import { screen, render, createWithWrapper, makeCustomStore } from "../test-utils";

type status = "checkIn" | "checkOut" | null;
describe("src/components/ProfileCard.tsx", () => {
  const profileCardProps: {
    handleFlip: any;
    handleCheckIn: any;
    handleCheckOut: any;
  } = {
    handleFlip: () => {},
    handleCheckIn: () => () => {},
    handleCheckOut: () => {},
  };

  beforeEach(() => {
    profileCardProps.handleCheckIn = jest.fn();
    profileCardProps.handleCheckOut = jest.fn();
    profileCardProps.handleFlip = jest.fn();
  });

  it("matches snapshot", () => {
    const component = createWithWrapper(<ProfileCard {...profileCardProps} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders userName and user image properly", () => {
    const user = { id: "sookang", profile: "someImage.jpg", state: "checkIn" as status };
    render(
      <ProfileCard {...profileCardProps} />,
      makeCustomStore({
        user: { ...userInitState, ...user },
      }),
    );
    const userNameTextEl = screen.queryByRole("heading", { name: user.id });
    const userImageEl = screen.getByRole("img", { name: /profile/i });

    expect(userNameTextEl).toBeInTheDocument();
    expect(userNameTextEl).toHaveTextContent(user.id);
    expect(userImageEl.getAttribute("src")).toBe(user.profile);

    const flipBtnEl = screen.getByTestId(/flip-button/i);
    // // 체크인시 사용
    // screen.getByText(/check in/i);
    // // checkout 시 사용
    // screen.getByRole("slider");
  });

  it("renders checkInUi when user's state is 'checkOut'", () => {
    const user = { id: "sookang", profile: "someImage.jpg", state: "checkOut" as status };

    render(
      <ProfileCard {...profileCardProps} />,
      makeCustomStore({
        user: { ...userInitState, ...user },
      }),
    );
    const checkInBtn = screen.queryByText(/check in/i);
    const checkOutBtn = screen.queryByRole("slider");

    expect(checkInBtn).toBeInTheDocument();
    expect(checkOutBtn).not.toBeInTheDocument();
  });

  it("renders checkInUi when user's state is 'null'", () => {
    const user = { id: "sookang", profile: "someImage.jpg", state: null as status };

    render(
      <ProfileCard {...profileCardProps} />,
      makeCustomStore({
        user: { ...userInitState, ...user },
      }),
    );
    const checkInBtn = screen.queryByText(/check in/i);
    const checkOutBtn = screen.queryByRole("slider");

    expect(checkInBtn).toBeInTheDocument();
    expect(checkOutBtn).not.toBeInTheDocument();
  });

  it("renders checkOutUi when user's state is 'checkIn'", () => {
    const user = { id: "sookang", profile: "someImage.jpg", state: "checkIn" as status };

    render(
      <ProfileCard {...profileCardProps} />,
      makeCustomStore({
        user: { ...userInitState, ...user },
      }),
    );
    const checkInBtn = screen.queryByText(/check in/i);
    const checkOutBtn = screen.queryByRole("slider");

    expect(checkInBtn).not.toBeInTheDocument();
    expect(checkOutBtn).toBeInTheDocument();
  });

  it("calls handleFlip when user clicks flip button", () => {
    const user = { id: "sookang", profile: "someImage.jpg", state: "checkIn" as status };
    render(
      <ProfileCard {...profileCardProps} />,
      makeCustomStore({
        user: { ...userInitState, ...user },
      }),
    );
    const flipBtnEl = screen.getByTestId(/flip-button/i);

    userEvent.click(flipBtnEl);

    expect(profileCardProps.handleFlip).toBeCalledTimes(1);
  });
});
