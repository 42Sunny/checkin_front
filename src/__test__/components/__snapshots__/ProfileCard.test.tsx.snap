// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/ProfileCard.tsx matches snapshot 1`] = `
<div
  className="profileCard"
>
  <img
    alt="logo"
    className="logo"
    src="42-logo-black.png"
  />
  <div
    className="util-box"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="flip-button"
      focusable="false"
      onClick={[MockFunction]}
      viewBox="0 0 24 24"
    >
      <path
        d="M3 13h2v-2H3v2zm0 4h2v-2H3v2zm0-8h2V7H3v2zm4 4h14v-2H7v2zm0 4h14v-2H7v2zM7 7v2h14V7H7z"
      />
    </svg>
  </div>
  <div
    className="profile-wrapper"
  >
    <img
      alt="profile"
      className="profile"
      src="user-default.png"
    />
    <h2
      className="userId"
    >
      
    </h2>
  </div>
  <hr
    className="divider"
  />
  <div
    className="box"
  >
    <p>
       클러스터 운영시간: 
      
    </p>
    <p>
       인포데스크 점심시간 
      11:00 ~ 12:00
    </p>
  </div>
  <form
    className="check-in-form"
  >
    <input
      className="cardNumber"
      inputMode="decimal"
      min={1}
      onChange={[Function]}
      placeholder="카드 번호"
      step={1}
      tabIndex={0}
      type="number"
      value=""
    />
    <div
      className="checkWrap"
    >
      <label
        className="text"
      >
        <input
          checked={false}
          className="checkbox"
          onChange={[Function]}
          type="checkbox"
        />
        방역수칙에 동의하고 입장합니다.
      </label>
      <button
        className="checkDetail"
        onClick={[Function]}
        type="button"
      >
        자세히
      </button>
    </div>
    <button
      className="button green"
      color="green"
      disabled={true}
      type="submit"
    >
      CHECK IN
    </button>
  </form>
</div>
`;
